listxattr: Test errno != 0 before returning error

fd0 opened the pull request:
Hi, and thanks for providing this library! I'm using it for [restic](https://github.com/restic/restic), a secure, fast and deduplicated backup program.

We've recently discovered that under some circumstances, `Listxattr()` on Darwin seems to return an error with `errno` set to zero, e.g. here: https://travis-ci.org/restic/restic/jobs/202245770#L138

Looking at the source code, the other functions check for `errno != 0` and only then return an error: https://github.com/pkg/xattr/blob/master/syscall_darwin.go#L13
It seems this check is missing for `listxattr()`.

A demo program can be found here: https://gist.github.com/11156f3d7cc5af48b9a5b654e872c2dc

For a file `foo` without extended attributes this prints the following:
```
$ go run main.go foo
[] &xattr.XAttrError{Op:"listxattr", Path:"foo", Name:"", Err:0x0}
```

You can see that an error is returned.

kuba-- commented:
@fd0 - Thanks Alex, unfortunately it's not enough. If we return from `listxattr` `(0, nil)` it will panic in `xattr. Listxattr` because of `size == 0` in https://github.com/pkg/xattr/blob/master/xattr_darwin.go#L29

But the fix generally looks trivial. In that case  `xattr. Listxattr` has to return `[]string{}, nil`

fd0 commented:
Hm, I added tests for `Listxattr()` and `Removexattr()`. These find the missing check for `syscall.Errno(0)`. I did not observe any panics, `make([]byte, 0)` does not panic on Go 1.7.5 or 1.8.

However I've also added an early return for that case.

Let me know if I need to change anything else.

I'd like to suggest running CI tests with Travis, would you consider enabling it and merging a PR which adds a `.travis.yml`?

kuba-- commented:
In `xattr_darwin.go` I would modify Listxattr:
```
func Listxattr(path string) ([]string, error) {
	// find size.
	size, err := listxattr(path, nil, 0, 0)
	if err != nil {
		return nil, &XAttrError{"listxattr", path, "", err}
	}

	if size > 0 {
		buf := make([]byte, size)
		// Read into buffer of that size.
		read, err := listxattr(path, &buf[0], size, 0)
		if err != nil {
			return nil, &XAttrError{"listxattr", path, "", err}
		}
		return nullTermToStrings(buf[:read]), nil
	}

	return []string{}, nil
}
```
and it may solve the problem on my mac

kuba-- commented:
@fd0 I would be happy to merge PR with `.travis.yml`

fd0 commented:
Cool, thanks!
